### PR TITLE
prov/verbs: Add partial support for the case where multiple verbs device...

### PR DESCRIPTION
**Pull request by a-ilango**
...s are present.

When multiple verbs devices are available on a system, construct verbs_info list
where each element corresponds to a device. If the node, service or hints provided
by the app uniquely identifies a device, check the hints against matched fi_info.

In fi_fabric, fi_domain and fi_endpint, search through verbs_info list to choose the
correct fi_info to check attributes against.

Initialize inject size from a user configurable default value instead of hints.

Signed-off-by: Arun C Ilango <arun.ilango@intel.com>

### CODE REVIEW

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
+	return 0;
+err:
+	fi_freeinfo(fi);
+	return ret;
+}
+
+static int fi_ibv_init_info(const struct fi_info *hints)
+{
+	struct ibv_context **ctx_list;
+	struct fi_info *fi, *fi_head, *fi_prev;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Please rename fi_prev to fi_tail, or just tail.

File: prov/verbs/src/fi_verbs.c [hunk 2]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
 	}
 
-	verbs_info = fi;
+	ctx_list = rdma_get_devices(&num_devices);
+	if (!num_devices) {
+		ret = (errno == ENODEV) ? -FI_ENODATA : -errno;
+		goto err1;
+	}
+
+	fi_head = fi_prev = fi = NULL;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: There's no need to set fi = NULL, we should check the return code to fi_ibv_get_info_ctx below

File: prov/verbs/src/fi_verbs.c [hunk 3]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
+	for (i = 0; i < num_devices; i++) {
+		fi_ibv_get_info_ctx(ctx_list[i], hints, &fi);
+		if (fi) {
+			if (!fi_head)
+				fi_head = fi;
+			if (fi_prev)
+				fi_prev->next = fi;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: This can be simplified to:
if (!head)
   head = fi;
else
   tail->next = fi;

File: prov/verbs/src/fi_verbs.c [hunk 4]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
+			fi_prev = fi;
+			fi = NULL;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: no need to set fi = NULL

File: prov/verbs/src/fi_verbs.c [hunk 5]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
+	ctx_list = rdma_get_devices(&num_devices);
+	if (!num_devices) {
+		ret = (errno == ENODEV) ? -FI_ENODATA : -errno;
+		goto err1;
+	}
+
+	fi_head = fi_prev = fi = NULL;
+	for (i = 0; i < num_devices; i++) {
+		fi_ibv_get_info_ctx(ctx_list[i], hints, &fi);
+		if (fi) {
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Check return code from ..get_info_ctx, rather than checking if fi has been set.

File: prov/verbs/src/fi_verbs.c [hunk 6]
```diff
@@ -943,24 +968,30 @@ static int fi_ibv_getinfo(uint32_t version, const char *node, const char *servic
...
 		return ret;
 
-	assert(!id->verbs || !strcmp(verbs_info->domain_attr->name,
-				     ibv_get_device_name(id->verbs->device)));
+	if (id->verbs) {
+		for (_info = verbs_info; _info; _info = _info->next) {
+			if (!strcmp(_info->domain_attr->name,
+				ibv_get_device_name(id->verbs->device)))
+				break;
+		}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Please move this into a function that searches for the correct verbs_info and returns it if found.

File: prov/verbs/src/fi_verbs.c [hunk 7]
```diff
@@ -943,24 +968,30 @@ static int fi_ibv_getinfo(uint32_t version, const char *node, const char *servic
...
+	} else {
+		_info = verbs_info;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: I wonder if there should be some generic info structure that isn't associated with a specific device.  Otherwise, the default settings are whatever the first device has.  This may be a separate change.

File: prov/verbs/src/fi_verbs.c [hunk 8]
```diff
@@ -943,24 +968,30 @@ static int fi_ibv_getinfo(uint32_t version, const char *node, const char *servic
 {
 	struct rdma_cm_id *id;
 	struct rdma_addrinfo *rai;
-	struct fi_info *fi;
+	struct fi_info *fi, *_info;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: The use of a leading underscore for a user variable name is discouraged.  This is often reserved by the compiler.

File: prov/verbs/src/fi_verbs.c [hunk 9]
```diff
@@ -3177,12 +3218,30 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
...
 	_domain = calloc(1, sizeof *_domain);
 	if (!_domain)
 		return -FI_ENOMEM;
 
+	for (_info = verbs_info; _info; _info = _info->next) {
+		if (!fi_ibv_check_domain_attr(info->domain_attr, _info)) {
+			match = 1;
+			break;
+		}
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Move into find match function

File: prov/verbs/src/fi_verbs.c [hunk 10]
```diff
@@ -3330,15 +3389,22 @@ static int fi_ibv_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric
...
 	if (ret)
 		return ret;
 
+	for (info = verbs_info; info; info = info->next) {
+		ret = fi_ibv_check_fabric_attr(attr, info);
+		if (!ret) {
+			match = 1;
+			break;
+		}
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Move into find match function

File: prov/verbs/src/fi_verbs.c [hunk 11]
```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
+			fi = NULL;
+		}
+	}
+
+	if (!fi_head) {
+		ret = -FI_ENODATA;
+		goto err2;
+	}
+
+	verbs_info = fi_head;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: The head variable can just be dropped in favor of using verbs_info directly, with verbs_info renamed to reflect that it is the head of a list.

File: prov/verbs/src/fi_verbs.c (hunk 7 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: But each of the fi_info structures in verbs_info would be associated with a device. The domain_attr->name would be set for each of them and all the attributes correspond to the device.

**Comment by a-ilango** [comment]
@shefty Thanks, I've updated the PR with suggested changes. I've left verbs_info variable name to remain the same for brevity.

File: prov/verbs/src/fi_verbs.c (hunk 7 above)
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: The verbs provider still supports listening across all underlying devices.  In that case, an fi_info structure would not be associated with a specific device.  Whether fi_info is associated with a device will depend on whether a source address has been selected/returned.

File: prov/verbs/src/fi_verbs.c [hunk 12]
```diff
@@ -938,29 +954,49 @@ static void fi_ibv_msg_ep_qp_init_attr(struct fi_ibv_msg_ep *ep,
 	}
 }
 
+static inline struct fi_info *fi_ibv_find_info(const char *fabric_name,
+		const char *domain_name)
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: remove 'inline' - let the compiler decide, which it will do anyway

File: prov/verbs/src/fi_verbs.c [hunk 13]
```diff
@@ -938,29 +954,49 @@ static void fi_ibv_msg_ep_qp_init_attr(struct fi_ibv_msg_ep *ep,
...
+{
+	struct fi_info *info;
+	int match = 0;
+
+	for (info = verbs_info; info; info = info->next) {
+		if ((domain_name && !strcmp(info->domain_attr->name, domain_name))
+			|| (fabric_name && !strcmp(info->fabric_attr->name, fabric_name))) {
+			match = 1;
+			break;
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: just return info here, no need for match variable, and return NULL below
nit: move the '||' to the end of the first line, rather than the start of the next

File: prov/verbs/src/fi_verbs.c [hunk 14]
```diff
@@ -938,29 +954,49 @@ static void fi_ibv_msg_ep_qp_init_attr(struct fi_ibv_msg_ep *ep,
...
 
-	assert(!id->verbs || !strcmp(verbs_info->domain_attr->name,
-				     ibv_get_device_name(id->verbs->device)));
+	if (id->verbs) {
+		check_info = fi_ibv_find_info(NULL, ibv_get_device_name(id->verbs->device));
 
-	if (!(fi = fi_dupinfo(verbs_info))) {
+	} else {
+		check_info = verbs_info;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: remove assert() above and add a check here:
if (!check_info)
   return error

File: prov/verbs/src/fi_verbs.c [hunk 15]
```diff
@@ -970,7 +1006,10 @@ static int fi_ibv_getinfo(uint32_t version, const char *node, const char *servic
 		goto err2;
 
 	if (hints) {
-		ret = fi_ibv_check_hints(hints, verbs_info);
+		/* TODO if we do not bind to a specific device we have to check 
+		 * check_info list and return those fi_info structures that match hints
+		 */
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Can you adjust the wording on this comment?  It's not clear what needs to be done.  I think this comment may be addressing my concerns expressed in another response regarding fi_info not associated with a device, but I'm not sure.

File: prov/verbs/src/fi_verbs.c [hunk 16]
```diff
@@ -2250,19 +2289,19 @@ fi_ibv_open_ep(struct fid_domain *domain, struct fi_info *info,
 		return -FI_EINVAL;
 
 	if (info->ep_attr) {
-		ret = fi_ibv_check_ep_attr(info->ep_attr, verbs_info);
+		ret = fi_ibv_check_ep_attr(info->ep_attr, _domain->info);
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Please rename variables that start with an underscore.  It looks like the name is already there, but go ahead and change it.

File: prov/verbs/src/fi_verbs.c [hunk 17]
```diff
@@ -3177,12 +3221,26 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
...
+		return ret;
+
 	_domain = calloc(1, sizeof *_domain);
 	if (!_domain)
 		return -FI_ENOMEM;
 
+	if (!(_domain->info = fi_dupinfo(fi))) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: What is the reason to make a copy of fi_info here?  This shouldn't be needed.  The user passes in fi_info into any relevant call.  If needed, the domain can store a pointer to the fi_info in the list, but even that shouldn't be needed.

File: prov/verbs/src/fi_verbs.c (hunk 16 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: OK. There are also other variables that start with underscore in many places. I shall do that cleanup in a separate PR.

File: prov/verbs/src/fi_verbs.c (hunk 14 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: I shall remove the assert and do away with the check. The exit condition for fi_ibv_init_info is verbs_info being set. I  guess that's sufficient.

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: > I shall remove the assert and do away with the check. The exit condition
> for fi_ibv_init_info is verbs_info being set. I guess that's sufficient.

I didn't quite follow this comment.  I was asking to add a check that check_info was not NULL after being set.

Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Sorry, I misread check_info as verbs_info in the previous comment.

File: prov/verbs/src/fi_verbs.c (hunk 17 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: The fi_info (fi) stored with domain here would be used to check the fi_info passed by the user in fi_endpoint call. 

fi is an element in verbs_info list. I'm making a copy of it so domain accesses only what it needs. fi_dupinfo copies just the element and not the entire list. This might be a little too much caution. Please let me know if that's not needed.

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: > The fi_info (fi) stored with domain here would be used to check the
> fi_info passed by the user in fi_endpoint call.

When fi_endpoint is called, it can either do a search for the fi_info again (we're talking about searching a list that usually contains 1 entry, maybe 2).  Or the domain can just store a pointer to the fi_info structure in the list.  There's no reason to make a copy.

File: prov/verbs/src/fi_verbs.c (hunk 15 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Quoting your response:
> The verbs provider still supports listening across all underlying devices.  In that case, an fi_info structure would not be associated with a specific device.  Whether fi_info is associated with a device will depend on whether a source address has been selected/returned.

Yes, this is what I'm trying to convey in the comment. Since the app doesn't bind to a specific device, I think we should return a list of fi_info structures to the user. Each of the fi_info will correspond to a device. The app would pass this along when calling fi_passive_ep. There's a TODO in fi_ibv_eq_cm_getinfo that says "This should copy the listening fi_info as the base". So the fi_info passed to passive_ep should be used as the base, right?

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Long term, this is what I think is needed:

If the app does not specify a source address, a list should be returned.  At least 1 entry in that list should _not_ be associated with a device.  The src_addr for that entry should contain a wildcard network address.  Entries that are associated with a specific device should contain a src_addr with a specific network address, but wildcard transport.

**Comment by a-ilango** [comment]
Updated the PR with suggested changes.

File: prov/verbs/src/fi_verbs.c [hunk 18]
```diff
@@ -815,25 +792,63 @@ static int fi_ibv_init_info(const struct fi_info *hints)
...
 	return 0;
-
-err3:
+err:
 	fi_freeinfo(fi);
-err2:
+	return ret;
+}
+
+static int fi_ibv_init_info(const struct fi_info *hints)
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: Something is not right here.  This call is initializing a list of fi_info structures.  The initialization should be completely independent of any 'hints'.

Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: The hints is passed along to fi_ibv_get_info_ctx -> fi_ibv_get_device_attrs -> fi_ibv_get_qp_cap to check if tx_attr->inject_size provided by the user is supported by the QP. The inject size is passed as ibv_qp_init_attr.max_inline_data. I'm not sure if there is another way to check the maximum inject size supported by the device.

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: > The hints is passed along to fi_ibv_get_info_ctx ->
> fi_ibv_get_device_attrs -> fi_ibv_get_qp_cap to check if tx_attr-
> >inject_size provided by the user is supported by the QP. The inject size
> is passed as ibv_qp_init_attr.max_inline_data. I'm not sure if there is
> another way to check the maximum inject size supported by the device.

I looked at that.  The problem is that the hints passed into the first fi_getinfo call ends up defining the inject size for all future calls.  That's not right.  This list should not be dependent on any value passed in by the user.

Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Ok. In that case, we can just use the default_inline_data variable to set the max inject size. It would be configurable by the user through a file similar to def_send_wr.

**Comment by a-ilango** [comment]
Now hints argument is removed from fi_ibv_init_info and other calls.

File: prov/verbs/src/fi_verbs.c [hunk 19]
```diff
@@ -938,29 +945,50 @@ static void fi_ibv_msg_ep_qp_init_attr(struct fi_ibv_msg_ep *ep,
...
-	 * check against if we have a valid ibv_context in id->verbs */
 	ret = fi_ibv_create_ep(node, service, flags, hints, &rai, &id);
 	if (ret)
 		return ret;
 
-	assert(!id->verbs || !strcmp(verbs_info->domain_attr->name,
-				     ibv_get_device_name(id->verbs->device)));
+	if (id->verbs) {
+		check_info = fi_ibv_find_info(NULL, ibv_get_device_name(id->verbs->device));
+
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: nit of a nit: extra blank line

File: prov/verbs/src/fi_verbs.c [hunk 20]
```diff
@@ -938,29 +945,50 @@ static void fi_ibv_msg_ep_qp_init_attr(struct fi_ibv_msg_ep *ep,
...
 }
 
+static struct fi_info *fi_ibv_find_info(const char *fabric_name,
+		const char *domain_name)
+{
+	struct fi_info *info;
+
+	for (info = verbs_info; info; info = info->next) {
+		if ((domain_name && !strcmp(info->domain_attr->name, domain_name)) ||
+			(fabric_name && !strcmp(info->fabric_attr->name, fabric_name))) {
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: last real change (no, really!) -- please change this check to something like:

if ((!domain || !strcmp(domain ...) && (!fabric || !strcmp(fabric ...))

I.e. if both the domain and fabric are provided, then both must match.  If neither are provider, we match on the item at the head of the list.

**Comment by a-ilango** [comment]
Updated the PR with suggested changes.
